File: muxer-commons/src/main/java/pt/um/tf/taskmux/commons/task/DummyTask.java
package pt.um.tf.taskmux.commons.task;

import pt.um.tf.taskmux.commons.error.MissingExecutorException;

import java.util.Random;
import java.util.concurrent.CompletableFuture;

public class DummyTask extends AsyncTask<Long> {
    private static long LOOPS_TO_DO = 16384;
    private CompletableFuture<Result<Long>> cp;
    private boolean started;

    public DummyTask(String id) {
        super(id);
        started = false;
    }

    protected DummyTask() {}

    @Override
    public CompletableFuture<Result<Long>> start() {
        if(!started) {
            started = true;
            startUpTask();
        }
        return cp;
    }

    private void startUpTask() {
        if(getExecutor() == null || getExecutor().isTerminated()) {
            cp = CompletableFuture.failedFuture(new MissingExecutorException());
        }
        else {
            cp = CompletableFuture.supplyAsync(this::dummyTask, getExecutor());
        }
    }

    private Result<Long> dummyTask() {
        var r = new Random(getURI().getPath().chars().asLongStream().sum());
        long id = 0;
        //Might intentionally divide by zero.
        //This simulates a possibly throwing long running background task.
        try{
            for(int i = 0; i < LOOPS_TO_DO; i++) {
                id += r.nextLong();
                id >>= 3;
                id *= 22;
                id += 2;
                id /= r.nextLong();
                try {
                    Thread.sleep(2, 800);
                } catch (InterruptedException e) {
                    //Silent night, Holy night.
                }
            }
            return new DummyResult(id);
        }
        catch (ArithmeticException e) {
            return new DummyResult(e);
        }
    }



    @Override
    public boolean completed() {
        return started && !cp.isCancelled() && cp.isDone();
    }

    @Override
    public void cancel() {
        if(started) {
            if (!cp.isDone()) {
                cp.cancel(true);
            }
        }
    }

    /*
    @Override
    public void writeObject(BufferOutput<?> buffer, Serializer serializer) {
        serializer.writeObject(getURI(), buffer).writeBoolean(started);
    }

    @Override
    public void readObject(BufferInput<?> buffer, Serializer serializer) {
        setURI(serializer.readObject(buffer));
        started = buffer.readBoolean();
        executorService = ForkJoinPool.commonPool();
    }

    @Override
    public void setExecutor(ExecutorService executorService) {
        this.executorService = executorService;
    }
    */
}

File: muxer-commons/src/main/kotlin/pt/um/tf/taskmux/commons/task/DummyTask.kt
package pt.um.tf.taskmux.commons.task

import pt.um.tf.taskmux.commons.URIGenerator
import pt.um.tf.taskmux.commons.error.MissingExecutorException

import java.util.Random
import java.util.concurrent.CompletableFuture
import java.util.function.Supplier

class DummyTask(u : URIGenerator = URIGenerator()) : AsyncTask<Long>(u) {
    private companion object {
       val LOOPS_TO_DO = 16384
    }
    private var cp : CompletableFuture<Result<Long>>? = null
    private var started : Boolean = false

    override fun start(): CompletableFuture<Result<Long>> {
        if(!started) {
            started = true
            startUpTask()
        }
        return cp as CompletableFuture<Result<Long>>
    }

    private fun startUpTask() {
        cp = if(executor == null || executor!!.isTerminated) {
            CompletableFuture.failedFuture(MissingExecutorException())
        }
        else {
            CompletableFuture.supplyAsync(dummyTask(), executor)
        }
    }

    private fun dummyTask() : Supplier<Result<Long>> {
        return Supplier {
            val r = Random(uri.path?.chars()?.asLongStream()?.sum() ?: 0)
            var id = 0L
            //Might intentionally divide by zero.
            //This simulates a possibly throwing long running background task.
            return@Supplier try{
                for(i in 0..LOOPS_TO_DO) {
                    id += r.nextInt()
                    id shl 3
                    id *= 22
                    id += 2
                    id /= r.nextInt()
                    Thread.sleep(2, 800)
                }
                DummyResult(id)
            } catch (e : Exception) {
                DummyResult(e = e)
            }
        }
    }


    override fun completed(): Boolean {
        return started && cp != null && !cp!!.isCancelled && !cp!!.isDone
    }

    override fun cancel() {
        if(started) {
            if (cp != null && !cp!!.isDone){
                cp?.cancel(true)
            }
        }
    }
}
